Implemented a toggle method on Item

rkoshak opened the pull request:
Signed-off-by: Richard Koshak <rlkoshak@gmail.com>

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").

ATTENTION: Don't use "git merge" when working with your pull request branch!
This can clutter your Git history and make your PR unuseable.
Use "git rebase" instead. See this forum post for further details:
https://community.openhab.org/t/rebase-your-code-or-how-to-fix-your-git-history-before-requesting-a-pull/129358

Add one or more appropriate labels to make your PR show up in the release notes, 
e.g. enhancement, bug, documentation, new feature.
This can only be done by yourself if you already contributed to this repo.

If your PR's code is not backward compatible with previous releases (which
should be avoided), add a message to the release notes by filing another PR:
https://github.com/openhab/openhab-distro/blob/main/distributions/openhab/src/main/resources/bin/update.lst
-->
# [Item] Added a Toggle Method to Item

# Description

Added a `toggle` method to Item which will toggle the state of the Item based on its current state. Any Item that supports OnOffType command or update will receive the opposite of its current `getStateAs(OnOffType)` as an update or command. Any Item that supports OpenClosedType updates will receive the opposite of its current `getStateOs(OpenClosedType)`. In all other cases an error is thrown (including attempting to toggle a Contact with a command). There is an argument `update` which, when provided with `true` will cause the new toggled state to be posted as an update. The default is `false` causing the new toggled state to be sent as a command.

There are many situations where one needs to toggle a binary type Item and this saves the user from needing to repeat that sort of logic over and over, similar to the already present `sendCommandIfDifferent()`.

# Testing

I tested most of the different permutations including:

- toggle on an Item whose state is UnDefType: generates error
- toggle on Switch without `update` argument: Item was commanded
- toggle on Switch with `update=false`: Item was commanded
- toggle on Switch with `update=true`: Item was updated
- toggle on Contact with `update=false`: generates error
- toggle on Contact with `update=true`: Item was updated

rkoshak commented:
Would it make sense to update/command the Item if its current state is `NULL` or `UNDEF` instead of throwing an error? It occurs to me that with the current error it might confuse users into thinking that the problem is the Item type and not that it's current state is the problem.

digitaldan commented on items/managed.js:
```diff
@@ -182,6 +182,28 @@ class Item {
+        else if(update && this.rawItem.getStateAs(OpenClosedType)) {
+            newState = (this.rawItem.getStateAs(OpenClosedType).toString() == 'CLOSED' ? 'OPEN' : 'CLOSED');
+        }
+        if(newState && update) this.postUpdate(newState);
+        else if(newState && !update) this.sendCommand(newState);
+        else throw Error('Items of type ' + this.rawItem.getType() + ' cannot be toggled via ' + ((update) ? 'an update' : 'a command'));
```
```suggestion
        if (newState) { 
          update ? this.postUpdate(newState) : this.sendCommand(newState);
        } else  { 
          throw Error('Items of type ' + this.rawItem.getType() + ' cannot be toggled via ' + ((update) ? 'an update' : 'a command'));
       }
```

digitaldan commented on items/managed.js:
```diff
@@ -182,6 +182,28 @@ class Item {
+        else if(update && this.rawItem.getStateAs(OpenClosedType)) {
+            newState = (this.rawItem.getStateAs(OpenClosedType).toString() == 'CLOSED' ? 'OPEN' : 'CLOSED');
+        }
+        if(newState && update) this.postUpdate(newState);
+        else if(newState && !update) this.sendCommand(newState);
+        else throw Error('Items of type ' + this.rawItem.getType() + ' cannot be toggled via ' + ((update) ? 'an update' : 'a command'));
```
```suggestion
        if (newState) { 
          update ? this.postUpdate(newState) : this.sendCommand(newState);
        } else  { 
          throw Error('Items of type ' + this.rawItem.getType() + ' cannot be toggled via ' + ((update) ? 'an update' : 'a command'));
       }
```

rkoshak replied:
Will do

digitaldan commented on items/managed.js:
```diff
@@ -182,6 +182,28 @@ class Item {
+     * @throws error if the Item cannot be toggled
+     */
+    toggle(update=false) {
+        var newState = null;
+        var oldState = this.rawItem.getStateAs(OnOffType);
```
```suggestion
        let newState = null;
        let oldState = this.rawItem.getStateAs(OnOffType);
```

minor nit :-)

digitaldan commented on items/managed.js:
```diff
@@ -182,6 +182,28 @@ class Item {
+        var newState = null;
+        var oldState = this.rawItem.getStateAs(OnOffType);
+        if(oldState) {
+            newState = (oldState.toString() == 'OFF') ? 'ON' : 'OFF';
```
```suggestion
            newState = oldState.toString() == 'OFF' ? 'ON' : 'OFF';
```

rkoshak replied:
Will do

digitaldan commented on items/managed.js:
```diff
@@ -182,6 +182,28 @@ class Item {
+            newState = (oldState.toString() == 'OFF') ? 'ON' : 'OFF';
+        }
+        else if(update && this.rawItem.getStateAs(OpenClosedType)) {
+            newState = (this.rawItem.getStateAs(OpenClosedType).toString() == 'CLOSED' ? 'OPEN' : 'CLOSED');
```
```suggestion
            newState = this.rawItem.getStateAs(OpenClosedType).toString() == 'CLOSED' ? 'OPEN' : 'CLOSED';
```

rkoshak replied:
Will do

digitaldan reviewed (changes requested):
I made a couple comments, one thing that seems odd to me about this is that we have one method which handles both a command and an update, which is kinda awkward to explain given some types support update and commands and some don't.  I get toggling switch like items and sending a command, rolling in updates and contacts seems to confuse this.  Maybe we should stick to what we support in rules as well ? 

 https://github.com/openhab/openhab-js/blob/d1b3cd5f87e967b59d8c612efc38bc83b7b3eb81/rules/operation-builder.js#L405

we could probably move this onto an item and reuse/reduce the amount of code needed.

rkoshak commented:
I think I may have accidentally rebaselined this PR. I thought I was on main on my fork. 

That wasn't my intention. I'm not sure how to proceed from here.

Clearly git and I do not get along. :-(

Anyway, I've addressed all the comments from above. I added a toggleCommand and toggleUpdate to Item and changed doToggle to call toggleCommand so it's all centralized.

digitaldan commented on rules/operation-builder.js:
```diff
@@ -404,21 +404,7 @@ class ToggleOperation extends OperationConfig {
      */
     doToggle() {
         let item = items.getItem(this.itemName);
```
Since we only use the reference once, might be cleaner to do
```suggestion
        items.getItem(this.itemName).toggleCommand();
```

rkoshak replied:
Done

digitaldan commented on time.js:
```diff
@@ -1,12 +1,28 @@
+ * @returns {number} duration from now to the ZonedDateTime in milliseconds
+ */
+time.ZonedDateTime.prototype.millisFromNow = function () {
+  return time.Duration.between(time.ZonedDateTime.now(), this).toMillis();
```
Not sure if this should be in this PR?

rkoshak replied:
Yes, I mucked things up and somehow merged upstream on this branch when I meant to on main instead. That's not actually my code. I'll try to get back to just my PR.

digitaldan commented on README.md:
```diff
@@ -144,19 +143,23 @@ The string representations of each of these objects are appended together in the
 see https://developer.mozilla.org/en-US/docs/Web/API/console for more information about console logging.
 
-### SetTimeout
+### Timers
```
I think this is from a another PR, as well as all the other time changes?

rkoshak replied:
See above

rkoshak replied:
Done

digitaldan reviewed (changes requested):
Thanks Richard, it looks like there are a lot of changes related to your PR around time, I also left a very small comment.  Can you  fix this so it just includes toggle functionality?  Might be easier to start fresh and cherry pick changes

digitaldan commented on items/managed.js:
```diff
@@ -181,6 +181,55 @@ class Item {
+     * command is sent).
+     * @throws error if the Item is uninitialized or a type that cannot be toggled or commanded
+     */
+    toggleCommand() {
```
```suggestion
    sendToggleCommand() {
```
wdyt?

rkoshak replied:
That works for me. I like consistency.

rkoshak replied:
Done

digitaldan commented on items/managed.js:
```diff
@@ -181,6 +181,55 @@ class Item {
+     * update is posted).
+     * @throws error if the Item is uninitialized or a type that cannot be toggled
+     */
+    toggleUpdate() {
```
```suggestion
    postToggleUpdate() {
```
wdyt?  I'm thinking we should try and be consistant since we use `postUpdate` elsewhere

rkoshak replied:
Same as above. I'll get a PR submitted later today with these two changes.

rkoshak replied:
Done

digitaldan reviewed (changes requested):
Thanks @rkoshak !  Looks pretty good to me, i only had 2 small comments about naming

digitaldan commented:
Hey @rkoshak any update on this?

rkoshak commented:
I've been trying to figure out how to service the unit tests using Mocha but haven't had a lot of time these last few weeks to do much of anything. I've but forgotten and am still working it. I can commit the charges without the tests to get this moving.

rkoshak commented:
I gave up on the uint testing. I might come back at a date with a new PR once I figure it out. Your requested changes have been made with this latest PR.

florian-h05 commented on rules/operation-builder.js:
```diff
@@ -403,22 +403,7 @@ class ToggleOperation extends OperationConfig {
-            default:
-                throw Error(`Toggle not supported for items of type ${item.type}`);
-        }
+        items.getItem(this.itemName).toggleCommand();
```
Wouldn't this be `sendToggleCommand` instead of `toggleCommand`?

rkoshak replied:
Fixed

florian-h05 reviewed (commented):
LGTM overall, but please have a look at my comment!

digitaldan commented:
@rkoshak i'm ok without the testing , if you want to fix the one issue mentioned i'm happy to merge, thanks!

rkoshak commented:
Should be good to go now.

florian-h05 reviewed (approved):
LGTM, thank you @rkoshak.
I did no testing but it should work.

digitaldan commented:
lgtm thanks!
